enterprise/server/routing/migration_operators/migration_operators.go

```diff
@@ -0,0 +1,136 @@
+package migration_operators
+
+import (
+	"context"
+	"io"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/batch_operator"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/buildbuddy-io/buildbuddy/server/util/usageutil"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+func ByteStreamCopy(ctx context.Context, router interfaces.CacheRoutingService, groupID string, b *batch_operator.DigestBatch) error {
+	ctx, cancel := context.WithCancel(usageutil.DisableUsageTracking(ctx))
+	defer cancel()
+
+	_, cas, err := router.GetCASClients(ctx)
+	if err != nil {
+		return err
+	}
+	res, err := cas.FindMissingBlobs(ctx, &repb.FindMissingBlobsRequest{
+		InstanceName:   b.InstanceName,
+		DigestFunction: b.DigestFunction,
+		BlobDigests:    b.Digests,
+	})
+	if err != nil {
+		return err
+	}
+	if len(res.MissingBlobDigests) == 0 {
+		// Nothing to copy!
+		return nil
+	}
+
+	primary, secondary, err := router.GetBSClients(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, d := range res.MissingBlobDigests {
+		if d.SizeBytes == 0 {
+			continue
+		}
+		// TODO(jdhollen): This should be using compression, when available.
+		r := digest.NewCASResourceName(d, b.InstanceName, b.DigestFunction)
+		readStream, err := primary.Read(ctx, &bspb.ReadRequest{ResourceName: r.DownloadString()})
+		if err != nil {
+			return err
+		}
+		writeClient, err := secondary.Write(ctx)
+		if err != nil {
+			return err
+		}
+		err = writeClient.Send(&bspb.WriteRequest{
+			ResourceName: r.NewUploadString(),
+			WriteOffset:  0,
```

Comment:
you'll need to make sure in the caller that these aren't offset reads right?

Reply:
the enqueued request is just based on a resource name.  if the resource isn't present, we've gotta copy the whole thing regardless.